### PR TITLE
Parse time zone info in Rust on Unix systems

### LINKED ISSUE #499: The call to `localtime_r` may be unsound
I found that `getenv` and `setenv` in libc are not thread-safe [1], and most impl of `localtime_r` in libc directly call `getenv` [2]. This means that `localtime_r` may have data race with `setenv`.

In order to ensure soundness of `setenv`, libstd add a lock to it [1], but this means that using `getenv` without libstd will be unsound. 

This problem is not easy to reproduce on glibc, because glibc's `localtime_r` caches timezone. but using musl can easily reproduce it.

1. https://github.com/rust-lang/rust/pull/24741
2. https://github.com/aosp-mirror/platform_bionic/blob/master/libc/tzcode/localtime.c#L1321 and https://git.musl-libc.org/cgit/musl/tree/src/time/__tz.c#n127

POC: https://gist.github.com/quininer/2063c31b0bc1753989122e782b182bea

### PR DESCRIPTION
**Pull request by djc**
Fixes #499 by vendoring a heavily modified copy of tz-rs.

Why we don't use tz-rs directly:

* tz-rs only supports 1.46+, whereas we get to support 1.32+
* Supply chain: the tz-rs maintainer is relatively unknown, and their GitHub profile is fairly opaque
* By constraining the capabilities to what chrono needs, the tz_info module is simpler than tz-rs

Alternative solutions: like time, we could check for concurrently running threads before running `localtime_r()`, leaving tzinfo handling to libc. Using well-tested Rust code to parse the timezone info files seems preferable to me over platform-specific hacks to find out if other threads are running.

@x-hgg-x thanks for your work on tz-rs!

### CODE REVIEW

**Comment by x-hgg-x** [comment]
For information, since the last release of `tz-rs`, I have added additional checks in the [`AlternateTime`](https://github.com/x-hgg-x/tz-rs/blob/ea3b698e5f93718a8b87371497a88130ef5cc77c/src/timezone/rule.rs#L325-L356) constructor, which are necessary for [this](https://github.com/x-hgg-x/tz-rs/blob/ea3b698e5f93718a8b87371497a88130ef5cc77c/src/timezone/rule.rs#L414) portion of code to work:
* Limit the UTC offset of the `LocalTimeType` fields between `-25h` and `26h`, which is the maximum allowed range when parsing a POSIX TZ string.
* [Check for DST transition rules consistency](https://github.com/x-hgg-x/tz-rs/blob/ea3b698e5f93718a8b87371497a88130ef5cc77c/src/timezone/rule.rs#L495): this ensures that the DST start and end time are always in the same order for each year. Otherwise, there will be an additional implementation-defined transition at the year boundary when the order is not the same between two consecutive years. This additional transition occurs at `Jan 1 00:00:00 UTC` for glibc and at `Jan 1 00:00:01 UTC` for musl libc, which is not what the TZ string would suggest.

I also added some [documentation](https://github.com/x-hgg-x/tz-rs/blob/ea3b698e5f93718a8b87371497a88130ef5cc77c/src/timezone/rule.rs#L108-L116) mentioning what a value of `Julian0WithLeap(365)` means for a non-leap year.

**Comment by djc** [comment]
> Any reason to use `once_cell` instead of [std](https://doc.rust-lang.org/std/index.html)::[sync](https://doc.rust-lang.org/std/sync/index.html)::[Once](https://doc.rust-lang.org/std/sync/struct.Once.html#)?

I wasn't aware of that! I found the interface substantially more painful but for this purpose it makes sense, so this PR now no longer pulls in once_cell. Thanks for the pointer!

> For information, since the last release of `tz-rs`, I have added additional checks in the [`AlternateTime`](https://github.com/x-hgg-x/tz-rs/blob/ea3b698e5f93718a8b87371497a88130ef5cc77c/src/timezone/rule.rs#L325-L356) constructor, which are necessary for [this](https://github.com/x-hgg-x/tz-rs/blob/ea3b698e5f93718a8b87371497a88130ef5cc77c/src/timezone/rule.rs#L414) portion of code to work:

Thanks, I will definitely be watching your follow-up commits after this is merged.

**Comment by esheppa** [comment]
Hi @djc - I've had an attempt to fix my above comments here and also added some more tests: https://github.com/esheppa/chrono/tree/tz-info-extra

(Please note that I've not yet fixed at least two areas with the *_from_local functions: `AlternativeTime::find_local_time_type_from_local` (local to and from UTC adjustments where necessary) and `TimeZoneRef::find_local_time_type_from_local` (leap second handling))

**Comment by djc** [comment]
@esheppa I'm not seeing your "above comments" -- did you forget to submit them? If you have a chance, could you submit any fixes as a PR that targets this PR's branch?

**Comment by esheppa** [comment]
I forgot to submit before, I've now submitted the comments and created PR  https://github.com/chronotope/chrono/pull/678

File: src/offset/local/unix.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+use crate::Utc;
+
+pub(super) fn now() -> DateTime<Local> {
+    let now = Utc::now();
+    DateTime::from_utc(now.naive_utc(), offset(now.timestamp()))
+}
+
+pub(super) fn naive_to_local(d: &NaiveDateTime, local: bool) -> DateTime<Local> {
+    let offset = match local {
+        true => offset(d.timestamp()),
```
[the remark this answers is not in the document]
Comment (x-hgg-x) [reply src/offset/local/unix.rs:24]: I just published the version `0.6.8` of `tz-rs` where I have added a method [`DateTime::find()`](https://github.com/x-hgg-x/tz-rs/blob/66285d60f3e5d3cad7c28f2bdcbd6ea3b66240a8/src/datetime/find.rs#L78) which covers this use case.

[the remark this answers is not in the document]
Comment (esheppa) [reply src/offset/local/unix.rs:24]: Nice @x-hgg-x  - I tried this out on my old [branch](https://github.com/esheppa/chrono/blob/1c5eb692258cb21f8c10d902bb8681c40bf29040/src/offset/local.rs#L65) and it passes all the tests locally (however I'd need to properly handle leap seconds)

[the remark this answers is not in the document]
Comment (x-hgg-x) [reply src/offset/local/unix.rs:24]: I see that you used [`unwrap()`](https://github.com/esheppa/chrono/blob/1c5eb692258cb21f8c10d902bb8681c40bf29040/src/offset/local.rs#L79-L80), which will panic if the datetime is after the last transition and there is no extra rule.

**Comment by djc** [comment]
@esheppa CI passed on #682, any clue why it failed after merging?

**Comment by esheppa** [comment]
@djc  - from what I can see #682 never ran on CI due to targeting this branch rather than main - I've temporarily switched #683 to compare with main so that CI can run there, then if that succeeds, I can change it back to target this branch

**Comment by esheppa** [comment]
@djc  - I've triggered the CI to test this at #689 however due to a rebase it can't be merged into this branch. I've created #690 which compresses all the localtime changes into one commit and targets this branch. I've temporarily edited the `test.yml` on #690 so that the CI runs.

**Comment by esheppa** [comment]
@djc - thanks for rebasing - I've now created #692 which includes just the changes to handle from_local_datetime. I'll try to avoid any more churn and just wait for any more rebases as required now

**Pull request merged**